### PR TITLE
MusicGen

### PR DESCRIPTION
**Pull request by barronalex**
Adds a [MusicGen](https://arxiv.org/abs/2306.05284) example.

Uses the existing T5 and Encodec examples for the text encoding / audio decoding.

I ended up removing the conversion step for T5 as in #1013 so we'll want to sequence that carefully if we merge this.

On M1 Max:

| Bench | Time (ms) |
| ------------- | ------------- |
| MLX  | 81.0 |
| Transformers PyTorch MPS  | 190.5 |

### CODE REVIEW

**Comment by awni** [comment]
Whoa cool!! Should we try an land 1013 ahead of this or is it simpler to close that and just land it here?

**Comment by angeloskath** [comment]
Music, images, text ... 🥳🥳🥳

I think this is more inline with the rest of the examples that have the `sanitize` pattern and I think @zcbenz agreed on the other PR.

I haven't done a thorough read but the only thing that may require some though is using the paths to import t5 and encodec. It is probably better than simply copying them into every project but I am not 100% on the paths. Otoh I don't know what other options we have:
- symlinks?
- top-level mlx-examples install?

**Comment by zcbenz** [comment]
Let's close #1013 in favor of this, it is easier rebasing some changes on this PR.

**Comment by barronalex** [comment]
Thanks @zcbenz, sorry for the concurrent PR!

Agreed on the ugliness of the paths. Symlinks sound pretty reasonable to me.

**Comment by awni** [comment]
Symlinks seem like a good call for now.

Though for the sake of discussion, I wonder  if we should consider packaging them up in some way. It's kind of common to use say a T5 in various places. It would be nice if I could do:

`pip install mlx_t5` and get the API to use it.

Or maybe we should package up a bunch of models here as like `mlx_models` and then you can do `from mlx_models import T5` or something like that.. not sure just brainstorming.

It's also not that bad to copy the code (which is one reason to keep it pretty minimal in the first place).

File: encodec/README.md [hunk 1]
```diff
@@ -33,13 +33,14 @@ An example using the model:
 
 ```python
 import mlx.core as mx
-from utils import load, load_audio, save_audio
+from encodec import EncodecModel
+from utils import load_audio, save_audio
 
 # Load the 48 KHz model and preprocessor.
-model, processor = load("mlx-community/encodec-48khz-float32")
+model, processor = EncodecModel.from_pretrained("mlx-community/encodec-48khz-float32")
```
Comment (awni) [inline encodec/README.md:40]: That's nice. Very 🤗

File: encodec/README.md [hunk 2]
```diff
@@ -33,13 +33,14 @@ An example using the model:
...
 
 # Load an audio file
-audio = load_audio("path/to/aduio", model.sampling_rate, model.channels)
+audio = load_audio("path/to/audio", model.sampling_rate, model.channels)
```
Comment (awni) [inline encodec/README.md:43]: ❤️

File: musicgen/generate.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+from utils import save_audio
+
+from musicgen import MusicGen
+
+
+def main(text: str, output_path: str, model_name: str, max_steps: int):
+    model = MusicGen.from_pretrained(model_name)
+    audio = model.generate(text, max_steps=max_steps)
+    save_audio(output_path, audio, model.sampling_rate)
```
Comment (awni) [inline musicgen/generate.py:13]: This might be the simplest generate script I've ever seen. Awesome!

**Review by awni — APPROVED** [review]
This is very nice! Good improvements all around.

Two requests:

1. Could you add an example-level README for this example? Can be pretty basic, just like a line about what the model is with some attribution and a line or two about how to run it.
2. Could you add a `requirements.txt` for this as well?

Can be pretty simple, just like

File: musicgen/musicgen.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+@partial(mx.compile, inputs=mx.random.state, outputs=mx.random.state)
+def top_k_sampling(
+    logits: mx.array, top_k: float, temperature: float, axis: int = -1
+) -> mx.array:
+    """
+    Apply top-k sampling to logits.
+
+    Args:
+        logits: The logits from the model's output.
+        top_k: The cumulative probability threshold for top-p filtering.
```
Comment (awni) [inline musicgen/musicgen.py:157]:
```suggestion
        top_k: The cumulative probability threshold for top-k filtering.
```

File: musicgen/musicgen.py [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+        temperature: Temperature parameter for softmax distribution reshaping.
+        axis: Axis along which to sample
+    Returns:
+        token selected based on the top-p criterion.
```
Comment (awni) [inline musicgen/musicgen.py:161]:
```suggestion
        token selected based on the top-k criterion.
```

File: musicgen/musicgen.py (hunk 4 above)
Comment (awni) [reply musicgen/musicgen.py:157]: Also I think that description of the `k` parameter is inaccurate.

File: musicgen/musicgen.py [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+            )
+
+        with open(path / "config.json", "r") as f:
+            config = SimpleNamespace(**json.load(f))
+            config.text_encoder = SimpleNamespace(**config.text_encoder)
+            config.audio_encoder = SimpleNamespace(**config.audio_encoder)
+            config.decoder = SimpleNamespace(**config.decoder)
+
+        weights = torch.load(path / "state_dict.bin", weights_only=True)["best_state"]
+        weights = {k: mx.array(v.numpy()) for k, v in weights.items()}
```
Comment (awni) [inline musicgen/musicgen.py:354]: I think it's better to go directly to `mx.array` here and not step through `numpy`. Avoids a copy and also won't crash for `bfloat16` (not sure if that's a thing)

File: musicgen/musicgen.py [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+
+        if cache is None:
+            cache = [None] * len(self.layers)
+
+        x = sum([self.emb[k](audio_tokens[..., k]) for k in range(self.num_codebooks)])
+
+        offset = cache[0].offset if cache[0] is not None else 0
+        offset = mx.full((audio_tokens.shape[0], 1, 1), offset)
+        pos_emb = create_sin_embedding(offset, self.hidden_size)
+        x += pos_emb
```
Comment (awni) [inline musicgen/musicgen.py:241]: II'm not sure if the models are ever in lower precision. If so a cast here will be important: `x += pos_emb.astype(x.dtype)`

File: musicgen/musicgen.py [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+        cache: list[KVCache] = None,
+    ):
+
+        if cache is None:
+            cache = [None] * len(self.layers)
+
+        x = sum([self.emb[k](audio_tokens[..., k]) for k in range(self.num_codebooks)])
+
+        offset = cache[0].offset if cache[0] is not None else 0
+        offset = mx.full((audio_tokens.shape[0], 1, 1), offset)
```
Comment (awni) [inline musicgen/musicgen.py:239]: I don't think you need to expand to the batch size here `mx.array(offset)` probably works ... though won't matter from a perf standpoint.

File: musicgen/musicgen.py (hunk 6 above)
Comment (barronalex) [reply musicgen/musicgen.py]: Really nice that that works!

**Comment by barronalex** [comment]
Thanks for the comments! I'd forgotten to `git add` the README and requirements they're in there now :)

File: musicgen/README.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+```
+# on macOS using Homebrew (https://brew.sh/)
+brew install ffmpeg
+```
+
+Install SciPy:
+
+```
+pip install scipy
+```
```
Comment (awni) [inline musicgen/README.md:28]: Nit: scipy is in the requirements.txt so you can probably remove the stuff here about installing scipy.

**Review by awni — APPROVED** [review]
Looks great! Let's 🚢

File: t5/convert.py [hunk 10]
```diff
@@ -1,75 +0,0 @@
-import numpy as np
```
[the remark this answers is not in the document]
Comment (barronalex) [reply t5/convert.py:1]: I think the 23 downloads are from your nice [open_sora](https://github.com/awni/mlx-examples/tree/open_sora/open_sora) example. That should still work OK since it has it's own T5 loading so we're probably good for the moment.
